src/libs/Violations/ViolationsUtils.ts

```diff
@@ -0,0 +1,64 @@
+import reject from 'lodash/reject';
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '@src/ONYXKEYS';
+import {PolicyCategories, PolicyTags, Transaction, TransactionViolation} from '@src/types/onyx';
+
+const ViolationsUtils = {
+    getViolationsOnyxData(
+        /** The transaction to check for policy violations. */
+        transaction: Transaction,
+        /** An array of existing transaction violations. */
+        transactionViolations: TransactionViolation[],
+        /** Indicates if the policy requires tags. */
+        policyRequiresTags: boolean,
+        /** Collection of policy tags and their enabled states. */
+        policyTags: PolicyTags,
+        /** Indicates if the policy requires categories. */
+        policyRequiresCategories: boolean,
+        /** Collection of policy categories and their enabled states. */
+        policyCategories: PolicyCategories,
+    ): {
+        onyxMethod: string;
+        key: string;
+        value: TransactionViolation[];
+    } {
+        let newTransactionViolations = [...transactionViolations];
+
+        if (policyRequiresCategories) {
+            const categoryViolationExists = transactionViolations.some((violation) => violation.name === 'categoryOutOfPolicy');
```

Comment:
Again, not the feedback that was sought here, but just to say it: if it were me, I would factor out the straight business logic into separate functions that do small things (e.g.., `constructCategoryViolations` for this section, `constructTagViolations` for the next, etc, and then call and combine all their results in this function).

I think that helps avoid some of the problems we've seen in other utility files with functions growing out of control, and becoming hard to refactor. But do with that what you wish.

Reply:
I tried that approach, but it felt much less readable when it was broken up. You had to jump all over the place to follow what was happening. 

Since the function is only ~40 lines, it felt ok to have it all in-line.